WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainStateHolder.kt

```diff
@@ -16,161 +16,215 @@
 package com.example.android.wearable.speaker
 
 import android.Manifest
-import android.app.Application
+import android.app.Activity
 import android.content.pm.PackageManager
 import android.media.AudioDeviceInfo
 import android.media.AudioManager
-import androidx.constraintlayout.motion.widget.MotionLayout
+import android.media.MediaPlayer
+import androidx.annotation.RequiresPermission
+import androidx.compose.foundation.MutatorMutex
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
 import androidx.core.content.ContextCompat
 import androidx.core.content.getSystemService
-import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.StateFlow
-import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.time.Duration
+import kotlin.coroutines.resume
 
 /**
- * A state holder driving the logic of [MainActivity].
+ * A state holder driving the logic of the app.
+ *
+ * This state holder is scoped to the composition (and therefore can be provided an [Activity]) to handle
+ * permission-related logic.
  */
 class MainStateHolder(
-    private val application: Application,
+    private val activity: Activity,
     private val requestPermission: () -> Unit,
     private val showPermissionRationale: () -> Unit,
     private val showSpeakerNotSupported: () -> Unit
 ) {
-
     /**
-     * The backing [MutableStateFlow] for the [appState].
+     * The [MutatorMutex] that guards the state of the app.
+     *
+     * Due to being a [MutatorMutex], this automatically handles cleanup of any ongoing asynchronous work,
+     * like playing music or recording, ensuring that only one operation is occurring at a time.
      */
-    private val appStateMutableStateFlow = MutableStateFlow<AppState>(AppState.Ready(transitionInstantly = true))
+    private val appStateMutatorMutex = MutatorMutex()
 
     /**
-     * A [StateFlow] of the primary app state.
+     * The primary app state.
      */
-    val appState = appStateMutableStateFlow.asStateFlow()
+    var appState by mutableStateOf<AppState>(AppState.Ready(transitionInstantly = true))
+        private set
 
     /**
-     * The backing [MutableStateFlow] for the [isPermissionDenied] flow.
+     * The progress of an ongoing recording.
      */
-    private val isPermissionDeniedMutableStateFlow = MutableStateFlow(false)
+    var recordingProgress by mutableStateOf(0f)
+        private set
 
     /**
-     * A [StateFlow] of whether we know the user has denied the record audio permission.
+     * `true` if we know the user has denied the record audio permission.
      */
-    val isPermissionDenied = isPermissionDeniedMutableStateFlow.asStateFlow()
+    var isPermissionDenied by mutableStateOf(false)
```

Comment:
Let's explain why this is not part of `AppState`

Reply:
I renamed `AppState` to `PlaybackState`, since the former implies that all of the state is contained in it, instead of having it separated out as appropriate.